### DISCUSSION TITLE
ERC-20 transactions (USDT/USDC etc)

### DISCUSSION THREAD

**Discussion by nikitaNotFound** [WalletConnect/walletconnect-monorepo#2413] [Developer Support] [answered]
It's a fork of https://github.com/orgs/WalletConnect/discussions/2382

My goal is connect user' wallet using WC v2 and Web3Modal and then request him for USDT transaction. Whether it'is possible?

Ether.js enables me to decode transaction data for calling USDT smart contract as it was suggested in the fork. Moreover I tested it in some wallets like TrustWallet and BitKeep. But push UI is totally not user friendly, so, I suppose I do something wrong. It just suggests user to pay for a gas without any information about token their transfer and it's amount. Here is an code example:

```
const usdtSmartContractAddress = "0xdac17f958d2ee523a2206206994597c13d831ec7";
const usdtContract = new ethers.Contract(usdtSmartContractAddress, UsdtAbi);
const amount = ethers.parseUnits(tokenAmount, 6);

const data = usdtContract.interface.encodeFunctionData('transfer', [recipientAddress, amount]);

const transaction = {
    from: recipientAddress,
    to: "0x7D0b77db2411A583c717e20f4F9824b5A65c0b81",
    value: null,
    data: data
};

const result = await signClient.request({
    topic: session.topic,
    request: {
         method: 'eth_sendTransaction',
         params: [transaction]
    },
    chainId: 'eip155:1'
});
```

Also I tried using EtheriumProvider with ether.js, but nothing happens at all:
```
const provider = await UniversalProvider.init({
   logger: "debug",
   projectId: projectId,
   metadata: {
       name: '...',
       description: 'Sample app to test WalletConnect.',
       url: '....',
       icons: ['....']
    },
    client: signClient
});
await provider.connect({
    namespaces
});

const web3Provider = new ethers.BrowserProvider(provider);

const usdtSmartContractAddress = "0xdac17f958d2ee523a2206206994597c13d831ec7";
const usdtContract = new ethers.Contract(usdtSmartContractAddress, UsdtAbi, web3Provider);
const amount = ethers.parseUnits(tokenAmount, 6);

const transferTx = await usdtContract.transfer('0x7D0b77db2411A583c717e20f4F9824b5A65c0b81', amount);

```
![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/46090847/f6638d35-6170-4e13-b02f-1ff602287e01)

**Comment by finessevanes** [comment, marked as answer]
In the first code block, it looks like you  might need to specify the USDT token contract address in the `to` property of the `transaction` object.

```js
const transaction = {
    from: recipientAddress,
    to: usdtSmartContractAddress,
    value: null,
    data: data
};
```

**Reply by nikitaNotFound** [reply]
> In the first code block, it looks like you might need to specify the USDT token contract address in the `to` property of the `transaction` object.
> 
> ```js
> const transaction = {
>     from: recipientAddress,
>     to: usdtSmartContractAddress,
>     value: null,
>     data: data
> };
> ```

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/46090847/166bef11-d7a3-4b4f-9ab4-519e9fd96134)

Thanks for your answer! I still get this bizarre modal in TrustWallet while making transaction. Do you know for sure whether it's possible to transfer USDT and other ERC-20 tokens using just smart contract instead of independent chain? I will be really appreciate for code sample then, because I fill with despair about it....

**Reply by finessevanes** [reply]
> I still get this bizarre modal in TrustWallet while making transaction. 

From what I'm seeing, the error is that you don't have enough ETH. You may need funds before transacting.

> Do you know for sure whether it's possible to transfer USDT and other ERC-20 tokens using just smart contract instead of independent chain?

Yep. It is. I found these resources for you online:
https://ethereum.org/en/developers/tutorials/transfers-and-approval-of-erc-20-tokens-from-a-solidity-smart-contract/
https://medium.com/coinmonks/hold-and-transfer-erc-20-on-a-smart-contract-1a640ca32e0d#:~:text=Solidity%20smart%20contracts%20can%20also,Releasing%20tokens%2C%20Banking%20and%20etc.&text=Let's%20look%20at%20how%20to,transfer%20them%20to%20an%20address.
